subsys/mgmt/mcumgr/grp/img_mgmt/src/img_mgmt_state.c

```diff
@@ -375,35 +412,80 @@ int img_mgmt_set_next_boot_slot(int slot, bool confirm)
 	const struct flash_area *fa;
 	int area_id = img_mgmt_flash_area_id(slot);
 	int rc = 0;
-	bool active = (slot == img_mgmt_active_slot(img_mgmt_active_image()));
-
-#if defined(CONFIG_MCUMGR_GRP_IMG_STATUS_HOOKS)
-	int32_t ret_rc;
-	uint16_t ret_group;
+	/* image the requested slot is defined within */
+	int image = img_mgmt_slot_to_image(slot);
+	/* active_slot is slot that is considered active/primary/executing
+	 * for a given image.
+	 */
+	int active_slot = img_mgmt_active_slot(image);
+	enum img_mgmt_next_boot_type type;
+	int next_boot_slot = img_mgmt_get_next_boot_slot(image, &type);
+
+	LOG_DBG("(%d, %s)", slot, confirm ? "confirm" : "test");
+	LOG_DBG("aimg = %d, img = %d, aslot = %d, slot = %d, nbs = %d",
+		img_mgmt_active_image(), image, active_slot, slot, next_boot_slot);
+
+	/* MCUmgr should not allow to confirm non-active image slots to prevent
+	 * confiming something that might not have been verified to actually be bootable
+	 * or have stuck in primary slot of other image. Unfortunately there was
+	 * a bug in logic that always allowed to confirm secondary slot of any
+	 * image. Now the behaviour is controlled via Kconfig options.
+	 */
+#ifndef CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_ANY
+	if (confirm && image != img_mgmt_active_image() &&
+	    (!IS_ENABLED(CONFIG_MCUMGR_GRP_IMG_ALLOW_CONFIRM_NON_ACTIVE_IMAGE_SECONDARY) ||
+	     slot == active_slot)) {
+		LOG_DBG("Not allowed to confirm non-active images");
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
+	}
 #endif
 
-	if (active) {
-		confirm = true;
-	} else {
-		if (slot != 1 &&
-		    !(CONFIG_MCUMGR_GRP_IMG_UPDATABLE_IMAGE_NUMBER == 2 && slot == 3)) {
-			return MGMT_ERR_EINVAL;
+	/* Setting test to active slot is not allowed. */
+	if (!confirm && slot == active_slot) {
+		return IMG_MGMT_RET_RC_INVALID_SLOT;
+	}
+
+	/* DO nothing when requested to test the non-active image already set for test. */
+	if (type == NEXT_BOOT_TYPE_TEST) {
+		return 0;
+	}
+
+	/* Normal boot means confirmed boot to either active slot or the opposite slot. */
+	if (type == NEXT_BOOT_TYPE_NORMAL) {
+		/* Do nothing when attempting to confirm slot that will be boot next time. */
+		if (confirm && slot == next_boot_slot) {
+			return 0;
 		}
+
+		/* Can not change slot once other than running has been confirmed. */
+		if (slot == active_slot && active_slot != next_boot_slot) {
+			return IMG_MGMT_RET_RC_IMAGE_ALREADY_PENDING;
+		}
+		/* Allow selecting non-active slot for boot */
```

Comment:
I wonder if we shouldn't disallow the following case too `!confirm && slot == next_boot_slot && active_slot != next_boot_slot` (to prevent trying to set test flag for the `next_boot_slot` when permanent swap is pending)